File: internal/cli/user_note.go
package cli

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"uni-context/internal/domain"
	"uni-context/internal/port"
	"uni-context/internal/service"

	"github.com/spf13/cobra"
)

var userCmd = &cobra.Command{
	Use:   "user",
	Short: "Manage personal-scope (user) knowledge",
}

var userNoteCmd = &cobra.Command{
	Use:   "note",
	Short: "Manage personal notes",
}

var (
	noteTitle      string
	noteTags       []string
	noteTagsFilter []string
	noteLimit      int
	noteFilePath   string
)

// userNoteLoadAppFn is the indirection that lets RunE tests swap in a
// stubbed *App without touching the real config/DB. Separate from
// embed.go's loadAppFn so each command file's tests are scoped to its
// own var. Defaults to the real loadApp in production.
var userNoteLoadAppFn = loadApp

var userNoteAddCmd = &cobra.Command{
	Use:   "add [content|-]",
	Short: "Add a personal note (positional arg, - for stdin, or --file <path>)",
	Long: `Add a personal note.

Content sources (mutually exclusive — pick one):
  positional arg    unictx user note add "hello world"
  -                 read from stdin (echo "hi" | unictx user note add -)
  --file <path>     import from a .txt or .md file (max 10 MB)

When --file is used without --title, the title defaults to the file's
basename with its extension stripped (weekly.md -> "weekly"). Markdown
files are tagged text/markdown so renderers can render them; other text
files default to text/plain.`,
	Args: cobra.MaximumNArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		// Rule 0: --file "" (explicit empty) must not fall through to
		// readContent(args), which would surface the misleading
		// "content required (positional arg or - for stdin)".
		if cmd.Flags().Changed("file") && noteFilePath == "" {
			return fmt.Errorf("--file: path cannot be empty")
		}

		var content string
		var mime string
		sourceMeta := map[string]any{}
		if noteFilePath != "" {
			// File import path.
			if len(args) > 0 { // Rule 1: mutual exclusion
				return fmt.Errorf("cannot combine --file with positional content or -")
			}
			if err := validateFileImport(noteFilePath); err != nil { // Rules 2-4
				return err
			}
			data, err := os.ReadFile(noteFilePath)
			if err != nil {
				return fmt.Errorf("read file: %w", err)
			}
			content = string(data)
			mime = mimeForTextFile(noteFilePath)
			if !cmd.Flags().Changed("title") {
				noteTitle = deriveDefaultTitle(noteFilePath)
			}
			sourceMeta["original_filename"] = filepath.Base(noteFilePath)
		} else {
			// Existing path: positional arg OR "-" stdin. Unchanged.
			c, err := readContent(args)
			if err != nil {
				return err
			}
			content = c
		}

		a, cfg, err := userNoteLoadAppFn()
		if err != nil {
			return err
		}
		defer a.Close()

		id, err := a.Ingest.Create(cmd.Context(), service.Input{
			Scope:       domain.ScopeUser,
			Kind:        domain.KindNote,
			Source:      domain.SourceManual,
			OwnerUserID: cfg.User.ID,
			Title:       noteTitle,
			Content:     content,
			Tags:        noteTags,
			MIME:        mime,
			SourceMeta:  sourceMeta,
		})
		if err != nil {
			return err
		}
		if flagJSON {
			printJSON(map[string]string{"id": id, "status": "added"})
		} else {
			fmt.Printf("added: %s\n", id)
		}
		return nil
	},
}

var userNoteListCmd = &cobra.Command{
	Use:   "list",
	Short: "List personal notes (newest first)",
	RunE: func(cmd *cobra.Command, args []string) error {
		a, cfg, err := loadApp()
		if err != nil {
			return err
		}
		defer a.Close()
		if noteLimit <= 0 {
			noteLimit = 20
		}
		items, _, err := a.Items.List(cmd.Context(), port.ItemFilter{
			Scopes:      []domain.Scope{domain.ScopeUser},
			OwnerUserID: cfg.User.ID,
			Kinds:       []domain.Kind{domain.KindNote},
			Tags:        noteTagsFilter,
			Limit:       noteLimit,
		})
		if err != nil {
			return err
		}
		if flagJSON {
			out := make([]map[string]any, 0, len(items))
			for _, it := range items {
				out = append(out, map[string]any{
					"id":         it.ID,
					"title":      it.Title,
					"tags":       it.Tags,
					"created_at": it.CreatedAt.Format(time.RFC3339),
				})
			}
			printJSON(out)
			return nil
		}
		if len(items) == 0 {
			fmt.Println("(no notes)")
			return nil
		}
		for _, it := range items {
			fmt.Println(formatListItem(it))
		}
		return nil
	},
}

var userNoteGetCmd = &cobra.Command{
	Use:   "get <id>",
	Short: "Show a single note",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		a, _, err := loadApp()
		if err != nil {
			return err
		}
		defer a.Close()
		item, err := a.Items.Get(cmd.Context(), args[0])
		if err != nil {
			return err
		}
		content := item.Content
		if flagJSON {
			printJSON(map[string]any{
				"id":         item.ID,
				"title":      item.Title,
				"summary":    item.Summary,
				"content":    content,
				"tags":       item.Tags,
				"created_at": item.CreatedAt.Format(time.RFC3339),
				"updated_at": item.UpdatedAt.Format(time.RFC3339),
			})
			return nil
		}
		fmt.Printf("id:    %s\n", item.ID)
		fmt.Printf("title: %s\n", item.Title)
		fmt.Printf("tags:  %s\n", strings.Join(item.Tags, ", "))
		fmt.Println("---")
		fmt.Println(content)
		return nil
	},
}

var userNoteDeleteCmd = &cobra.Command{
	Use:   "delete <id>",
	Short: "Delete a note",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		a, _, err := loadApp()
		if err != nil {
			return err
		}
		defer a.Close()
		if err := a.Items.Delete(cmd.Context(), args[0]); err != nil {
			return err
		}
		if flagJSON {
			printJSON(map[string]string{"id": args[0], "status": "deleted"})
		} else {
			fmt.Printf("deleted: %s\n", args[0])
		}
		return nil
	},
}

func init() {
	userNoteAddCmd.Flags().StringVar(&noteTitle, "title", "", "note title")
	userNoteAddCmd.Flags().StringSliceVar(&noteTags, "tag", nil, "tags (comma-separated or repeat)")
	userNoteAddCmd.Flags().StringVar(&noteFilePath, "file", "", "import content from a file (text only)")
	userNoteListCmd.Flags().StringSliceVar(&noteTagsFilter, "tag", nil, "filter by tag (OR semantics; comma-separated or repeat)")
	userNoteListCmd.Flags().IntVar(&noteLimit, "limit", 20, "max items to return")

	userNoteCmd.AddCommand(userNoteAddCmd, userNoteListCmd, userNoteGetCmd, userNoteDeleteCmd)
	userCmd.AddCommand(userNoteCmd)
	rootCmd.AddCommand(userCmd)
}

func readContent(args []string) (string, error) {
	if len(args) == 0 || args[0] != "-" {
		if len(args) == 0 {
			return "", fmt.Errorf("content required (positional arg or - for stdin)")
		}
		return args[0], nil
	}
	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		return "", fmt.Errorf("read stdin: %w", err)
	}
	return string(data), nil
}

func inputFromFlags(scope domain.Scope, kind domain.Kind, source domain.Source,
	owner, project, title, content string, tags []string,
) service.Input {
	return service.Input{
		Scope:       scope,
		Kind:        kind,
		Source:      source,
		OwnerUserID: owner,
		ProjectID:   project,
		Title:       title,
		Content:     content,
		Tags:        tags,
	}
}

// listPreviewLen is the maximum rune count of content shown as a fallback
// title in list output when an item has no title.
const listPreviewLen = 50

// formatListItem renders one row of `user note list`. When the item has a
// non-empty title, the title is shown verbatim. When the title is empty
// (common when `add` was called without --title), a preview of the inline
// content is shown instead so the user sees something useful. When content
// is also empty (externalized to FileStore), an "(externalized)" placeholder
// is shown — the full content can always be retrieved with `get <id>`.
func formatListItem(item domain.ContextItem) string {
	label := item.Title
	if label == "" {
		switch {
		case item.Content != "":
			label = previewRunes(item.Content, listPreviewLen)
		case item.ContentURI != "":
			label = "(externalized)"
		default:
			label = "(no content)"
		}
	}
	tags := strings.Join(item.Tags, ",")
	return fmt.Sprintf("%s  %s  [%s]", item.ID, label, tags)
}

// previewRunes returns the first n runes of s, appending an ellipsis if s
// was truncated.
func previewRunes(s string, n int) string {
	runes := []rune(s)
	if len(runes) <= n {
		return s
	}
	return string(runes[:n]) + "…"
}

// maxFileBytes is the file import size cap. Enforced via os.Stat before
// os.ReadFile so a rejected file never allocates a buffer. 10 MB is a
// guardrail against accidentally loading huge files, not a security boundary.
const maxFileBytes int64 = 10 * 1024 * 1024

// mimeForTextFile maps a small set of text file extensions to MIME types.
// Unknown extensions default to text/plain — binary support is out of scope.
// Case-insensitive via strings.ToLower so weekly.MD is still markdown.
// Adding new text types later (.org, .rst) is a one-liner here.
func mimeForTextFile(path string) string {
	switch strings.ToLower(filepath.Ext(path)) {
	case ".md", ".markdown":
		return "text/markdown"
	default:
		return "text/plain"
	}
}

// deriveDefaultTitle extracts a human-friendly title from a file path by
// taking the basename and stripping the last extension. Used when the user
// runs `--file weekly.md` without `--title`. Only the last extension is
// stripped (archive.tar.gz → "archive.tar") to match user intuition.
// A leading-dot file (.bashrc) keeps its full basename (dot at index 0
// is not stripped).
func deriveDefaultTitle(path string) string {
	base := filepath.Base(path)
	if dot := strings.LastIndex(base, "."); dot > 0 {
		base = base[:dot]
	}
	return base
}

// checkFileSize is a pure function so tests can sweep synthetic sizes
// (0, at-cap, cap+1) without writing real fixtures to disk.
func checkFileSize(size int64) error {
	if size > maxFileBytes {
		return fmt.Errorf("file too large: %d bytes (max %d)", size, maxFileBytes)
	}
	return nil
}

// validateFileImport runs the file-level validation rules (Rules 2-4 from
// the spec): file must exist, be a regular file, and be within the size cap.
// os.Stat runs before any os.ReadFile so oversized files are rejected
// without allocating a buffer. Rule 0 (empty path) and Rule 1 (mutual
// exclusion with positional args) are handled in RunE before this helper.
func validateFileImport(path string) error {
	info, err := os.Stat(path)
	if err != nil {
		return fmt.Errorf("stat file: %w", err)
	}
	if !info.Mode().IsRegular() {
		return fmt.Errorf("not a regular file: %s", path)
	}
	if err := checkFileSize(info.Size()); err != nil {
		return err
	}
	return nil
}

File: python/src/unictx/cli/user_note.py
"""``user note`` subcommands — add / get / list / delete.

Faithful port of Go's ``internal/cli/user_note.go``. Behaviors preserved:

  - **Rule 0:** ``--file ""`` (explicit empty) must NOT fall through to
    ``read_content`` (which would surface the misleading "content
    required" error). Surface a clear ``--file: path cannot be empty``.
  - **Rule 1:** ``--file`` and positional content are mutually exclusive.
  - **Rules 2-4:** file must exist, be a regular file, and be within
    the size cap. Validation runs before any read so oversized files
    are rejected without allocating a buffer.
  - **--engine** validated up front (typo feedback without IO wait).
    Acceptable values: ``fitz`` (PyMuPDF, default), ``shell``, ``http``.
    Mirrors Go's ``gxpdf|shell|http`` trio — ``fitz`` replaces ``gxpdf``
    because the Python port uses PyMuPDF, not gxpdf.
  - **--file title default:** when ``--title`` not set, derive from
    filename (last extension stripped; dotfiles keep full basename).
  - **Output modes:** ``--json`` (global flag) emits machine-readable
    JSON; otherwise plain-text human output.

Typer composition
=================

The root ``app`` lives in :mod:`unictx.cli.app`. This module builds two
nested Typers::

    user_app (Typer)            → unictx user ...
        note_app (Typer)        → unictx user note ...
            add / get / list / delete

The root app registers ``user_app`` via ``app.add_typer(user_app, name="user")``
in :mod:`unictx.cli.__init__` (Tasks 6.2-6.5 register their own Typers
the same way).

Container injection
===================

Production calls :func:`_default_load_container` which composes
``wire(load_config(get_config_path()))`` once per invocation. Tests
monkeypatch ``_load_container`` to return a tmp_path-backed container
(Go's ``userNoteLoadAppFn`` pattern).
"""

from __future__ import annotations

import os
import sys
from pathlib import Path
from typing import Any

import typer

from unictx.cli.app import AppContainer, get_config_path, wire
from unictx.config import load as load_config
from unictx.items.ingest import Input
from unictx.items.models import Kind, Scope, Source
from unictx.items.repo import ItemFilter
from unictx.pdf.factory import build_extractor_for_engine

__all__ = [
    "check_file_size",
    "derive_default_title",
    "format_list_item",
    "mime_for_file",
    "note_app",
    "preview_runes",
    "user_app",
    "validate_file_import",
]


# ---------------------------------------------------------------------------
# Typer sub-apps
# ---------------------------------------------------------------------------


user_app = typer.Typer(
    name="user",
    help="Manage personal-scope (user) knowledge.",
    no_args_is_help=True,
    add_completion=False,
)

note_app = typer.Typer(
    name="note",
    help="Manage personal notes.",
    no_args_is_help=True,
    add_completion=False,
)

user_app.add_typer(note_app, name="note")


# ---------------------------------------------------------------------------
# Container injection
# ---------------------------------------------------------------------------


def _default_load_container() -> AppContainer:
    """Production: load config from --config path (or XDG default) + wire."""
    return wire(load_config(get_config_path()))


# Module-level seam — tests monkeypatch this. Defaults to production path.
_load_container = _default_load_container


# ---------------------------------------------------------------------------
# Helpers — pure functions (unit-tested in test_user_note.py)
# ---------------------------------------------------------------------------


# File import size cap. Mirrors Go's maxFileBytes. Bumped from 10MB → 50MB
# for PDF support (academic papers commonly 5-15MB, scanned textbooks
# 20-80MB). Text files rarely approach this; cap exists as a guardrail.
_MAX_FILE_BYTES = 50 * 1024 * 1024

# Max rune count of content shown as a fallback title in list output.
_LIST_PREVIEW_LEN = 50

# Acceptable --engine values. Mirrors Go's gxpdf|shell|http trio — "fitz"
# replaces "gxpdf" because the Python port uses PyMuPDF, not gxpdf.
_VALID_ENGINES = {"fitz", "shell", "http"}


def mime_for_file(path: str) -> str:
    """Return MIME type for a file based on its extension.

    Renamed from Go's ``mimeForTextFile`` when PDF support was added
    (the old name lied once it returned application/pdf). Unknown
    extensions fall back to ``text/plain`` (backward compat for users
    who pass weirdly-named text files).
    """
    ext = Path(path).suffix.lower()
    if ext in {".md", ".markdown"}:
        return "text/markdown"
    if ext == ".pdf":
        return "application/pdf"
    return "text/plain"


def derive_default_title(path: str) -> str:
    """Extract a human-friendly title from a file path.

    Takes the basename and strips the last extension. Used when the user
    runs ``--file weekly.md`` without ``--title``. Only the LAST
    extension is stripped (``archive.tar.gz`` → ``archive.tar``) to
    match user intuition. A leading-dot file (``.bashrc``) keeps its
    full basename (dot at index 0 is not stripped).
    """
    base = Path(path).name
    dot = base.rfind(".")
    if dot > 0:
        base = base[:dot]
    return base


def check_file_size(size: int) -> str | None:
    """Return error message if ``size`` exceeds cap, else None."""
    if size > _MAX_FILE_BYTES:
        return f"file too large: {size} bytes (max {_MAX_FILE_BYTES})"
    return None


def validate_file_import(path: str) -> str | None:
    """Run file-level validation rules (Rules 2-4 from the spec).

    - File must exist (os.stat error surfaces with context).
    - File must be a regular file (not a directory, device, socket).
    - File must be within the size cap.

    Returns error message string on failure, ``None`` on success.
    """
    try:
        st = os.stat(path)
    except OSError as exc:
        return f"stat file: {exc}"
    if not os.path.isfile(path):
        return f"not a regular file: {path}"
    if err := check_file_size(st.st_size):
        return err
    return None


def preview_runes(s: str, n: int) -> str:
    """Return first ``n`` runes of ``s`` with ellipsis if truncated."""
    runes = list(s)
    if len(runes) <= n:
        return s
    return "".join(runes[:n]) + "…"


def format_list_item(item: Any) -> str:
    """Render one row of ``user note list``.

    When the item has a non-empty title, the title is shown verbatim.
    When the title is empty (common when ``add`` was called without
    ``--title``), a preview of the inline content is shown instead so
    the user sees something useful. When content is also empty
    (externalized to FileStore), an ``(externalized)`` placeholder is
    shown — the full content can always be retrieved with ``get <id>``.
    """
    label = item.title
    if not label:
        if item.content:
            label = preview_runes(item.content, _LIST_PREVIEW_LEN)
        elif item.content_uri:
            label = "(externalized)"
        else:
            label = "(no content)"
    tags = ",".join(item.tags)
    return f"{item.id}  {label}  [{tags}]"


def _read_content(content_arg: str | None) -> tuple[str, str | None]:
    """Resolve the positional-or-stdin content source.

    Returns ``(content, error_msg)``. On success ``error_msg`` is None.
    Mirrors Go's ``readContent``:
      - No positional → error "content required".
      - ``-`` → read stdin.
      - Anything else → use the positional verbatim.
    """
    if content_arg is None or content_arg == "":
        return "", "content required (positional arg or - for stdin)"
    if content_arg != "-":
        return content_arg, None
    # stdin
    try:
        data = sys.stdin.read()
    except OSError as exc:
        return "", f"read stdin: {exc}"
    return data, None


# ---------------------------------------------------------------------------
# Commands
# ---------------------------------------------------------------------------


@note_app.command("add")
def add(
    content: str | None = typer.Argument(
        None,
        help='Note content (positional, "-" for stdin, or omit with --file).',
    ),
    title: str = typer.Option("", "--title", help="Note title."),
    tag: list[str] = typer.Option([], "--tag", help="Tags (repeatable: --tag red --tag blue)."),
    file: str = typer.Option("", "--file", help="Import content from a file (.txt, .md, .pdf)."),
    engine: str = typer.Option(
        "",
        "--engine",
        help='PDF extractor override: "fitz", "shell", or "http". Empty uses the config default.',
    ),
) -> None:
    """Add a personal note (positional arg, - for stdin, or --file <path>)."""
    from unictx.cli.app import is_json_mode

    # Rule 0: --file "" (explicit empty) must surface a clear error,
    # not the misleading "content required" from _read_content.
    # Typer doesn't distinguish "not passed" from "passed empty" by
    # default; we use a sentinel by checking file == "" after Typer
    # parsing. The mutual-exclusion + content-required checks below
    # handle the rest. (Go uses cobra.Flags().Changed("file").)
    if file == "" and content is None:
        # Bare `add` with no positional + no --file → content-required error.
        typer.echo("content required (positional arg or - for stdin)", err=True)
        raise typer.Exit(code=2)

    # Engine validation runs BEFORE any IO so users get typo feedback
    # instantly, without waiting for a file read or stat.
    if engine and engine not in _VALID_ENGINES:
        typer.echo(
            f'unknown pdf engine "{engine}" (want {"|".join(sorted(_VALID_ENGINES))})',
            err=True,
        )
        raise typer.Exit(code=2)

    note_content: str
    mime: str = ""
    source_meta: dict[str, Any] = {}

    if file:
        # File import path.
        if content is not None:
            # Rule 1: mutual exclusion.
            typer.echo(
                "cannot combine --file with positional content or -",
                err=True,
            )
            raise typer.Exit(code=2)
        if err := validate_file_import(file):
            typer.echo(err, err=True)
            raise typer.Exit(code=2)
        try:
            data = Path(file).read_bytes()
        except OSError as exc:
            typer.echo(f"read file: {exc}", err=True)
            raise typer.Exit(code=2) from exc
        note_content = data.decode("utf-8", errors="replace")
        mime = mime_for_file(file)
        if not title:
            title = derive_default_title(file)
        source_meta["original_filename"] = Path(file).name
    else:
        # Positional / stdin path.
        c, err = _read_content(content)
        if err:
            typer.echo(err, err=True)
            raise typer.Exit(code=2)
        note_content = c

    container = _load_container()
    try:
        # Engine override: when --engine is set OR the file is a PDF,
        # build an extractor explicitly. The constructor default from
        # wire() only fires for non-PDF-aware callers — the CLI always
        # takes this path for PDFs so the choice is per-invocation.
        extractor_override = None
        if engine or mime == "application/pdf":
            engine_name = engine or container.config.pdf.engine
            if not engine_name:
                typer.echo(
                    "pdf extraction not configured: set pdf.engine in config or pass --engine",
                    err=True,
                )
                raise typer.Exit(code=2)
            try:
                extractor_override = build_extractor_for_engine(engine_name, container.config.pdf)
            except Exception as exc:
                typer.echo(f"build pdf extractor: {exc}", err=True)
                raise typer.Exit(code=2) from exc

        ingest_input = Input(
            scope=Scope.USER,
            kind=Kind.NOTE,
            source=Source.MANUAL,
            owner_user_id=container.config.user.id,
            title=title,
            content=note_content,
            tags=list(tag),
            mime=mime,
            source_meta=source_meta,
        )
        try:
            new_id = container.ingest.create(ingest_input, extractor=extractor_override)
        except Exception as exc:
            typer.echo(str(exc), err=True)
            raise typer.Exit(code=1) from exc

        if is_json_mode():
            from unictx.cli.output import print_json

            print_json({"id": new_id, "status": "added"})
        else:
            typer.echo(f"added: {new_id}")
    finally:
        container.close()


@note_app.command("get")
def get(
    item_id: str = typer.Argument(..., help="Note id."),
) -> None:
    """Show a single note (full content, hydrated from FileStore if externalized)."""
    from unictx.cli.app import is_json_mode

    container = _load_container()
    try:
        try:
            item = container.items.get(item_id)
        except Exception as exc:
            typer.echo(str(exc), err=True)
            raise typer.Exit(code=1) from exc

        if is_json_mode():
            from unictx.cli.output import print_json

            print_json(
                {
                    "id": item.id,
                    "title": item.title,
                    "summary": item.summary,
                    "content": item.content,
                    "tags": item.tags,
                    "created_at": item.created_at,
                    "updated_at": item.updated_at,
                }
            )
            return

        typer.echo(f"id:    {item.id}")
        typer.echo(f"title: {item.title}")
        typer.echo(f"tags:  {', '.join(item.tags)}")
        typer.echo("---")
        typer.echo(item.content)
    finally:
        container.close()


@note_app.command("list")
def list_notes(
    tag: list[str] = typer.Option([], "--tag", help="Filter by tag (OR semantics; repeatable)."),
    limit: int = typer.Option(20, "--limit", help="Max items to return (default 20)."),
) -> None:
    """List personal notes (newest first)."""
    from unictx.cli.app import is_json_mode

    if limit <= 0:
        limit = 20

    container = _load_container()
    try:
        filter_ = ItemFilter(
            scopes=[Scope.USER],
            owner_user_id=container.config.user.id,
            kinds=[Kind.NOTE],
            tags=list(tag),
            limit=limit,
        )
        try:
            items, _ = container.items.list(filter_)
        except Exception as exc:
            typer.echo(str(exc), err=True)
            raise typer.Exit(code=1) from exc

        if is_json_mode():
            from unictx.cli.output import print_json

            print_json(
                [
                    {
                        "id": it.id,
                        "title": it.title,
                        "tags": it.tags,
                        "created_at": it.created_at,
                    }
                    for it in items
                ]
            )
            return

        if not items:
            typer.echo("(no notes)")
            return
        for it in items:
            typer.echo(format_list_item(it))
    finally:
        container.close()


@note_app.command("delete")
def delete(
    item_id: str = typer.Argument(..., help="Note id."),
) -> None:
    """Delete a note."""
    from unictx.cli.app import is_json_mode

    container = _load_container()
    try:
        try:
            container.items.delete(item_id)
        except Exception as exc:
            typer.echo(str(exc), err=True)
            raise typer.Exit(code=1) from exc

        if is_json_mode():
            from unictx.cli.output import print_json

            print_json({"id": item_id, "status": "deleted"})
        else:
            typer.echo(f"deleted: {item_id}")
    finally:
        container.close()
